File: MyApp.Client/src/components/css.ts
import { TableStyle, TableStyleOptions } from "./types"

function hasTableStyle(style:TableStyleOptions, target:TableStyle) {
    return Array.isArray(style)
        ? style.indexOf(target) >= 0
        : style == target || style.includes(target)
}

export const grid = {
    getGridClass(style:TableStyleOptions="stripedRows") { return grid.gridClass },
    getGrid2Class(style:TableStyleOptions="stripedRows") { return hasTableStyle(style,"fullWidth")
        ? "overflow-x-auto" 
        : grid.grid2Class },
    getGrid3Class(style:TableStyleOptions="stripedRows") { return hasTableStyle(style,"fullWidth")
        ? "inline-block min-w-full py-2 align-middle"
        : grid.grid3Class },
    getGrid4Class(style:TableStyleOptions="stripedRows") { return hasTableStyle(style,"whiteBackground")
        ? ""
        : hasTableStyle(style,"fullWidth")
            ? "overflow-hidden shadow-sm ring-1 ring-black ring-opacity-5"
            : grid.grid4Class },
    getTableClass(style:TableStyleOptions="stripedRows") { return hasTableStyle(style,"fullWidth") || hasTableStyle(style,"verticalLines")
        ? "min-w-full divide-y divide-gray-300"
        : grid.tableClass },
    getTheadClass(style:TableStyleOptions="stripedRows") { return hasTableStyle(style,"whiteBackground")
        ? ""
        : grid.theadClass },
    getTheadRowClass(style:TableStyleOptions="stripedRows") { 
        return grid.theadRowClass + (hasTableStyle(style,"verticalLines") ? " divide-x divide-gray-200 dark:divide-gray-700" : "") },
    getTheadCellClass(style:TableStyleOptions="stripedRows") { 
        return grid.theadCellClass + (hasTableStyle(style,"uppercaseHeadings") ? " uppercase" : "") },
    getTbodyClass(style:TableStyleOptions="stripedRows") {
        return (hasTableStyle(style,"whiteBackground") || hasTableStyle(style,"verticalLines")
            ? "divide-y divide-gray-200 dark:divide-gray-800"
            : grid.tableClass)
        + (hasTableStyle(style,"verticalLines") ? " bg-white" : "") },
    getTableRowClass(style:TableStyleOptions="stripedRows", i:number, selected:boolean, allowSelection:boolean) {
        return (allowSelection ? "cursor-pointer " : "") + 
            (selected ? "bg-indigo-100 dark:bg-blue-800" : (allowSelection ? "hover:bg-yellow-50 dark:hover:bg-blue-900 " : "") + (hasTableStyle(style,"stripedRows")
                ? (i % 2 == 0 ? "bg-white dark:bg-black" : "bg-gray-50 dark:bg-gray-800")
                : "bg-white dark:bg-black")) + 
             (hasTableStyle(style,"verticalLines") ? " divide-x divide-gray-200 dark:divide-gray-700" : "")
    },

    gridClass: "flex flex-col",
    //original -margins + padding forces scroll bars when parent is w-full for no clear benefits?
    //original: -my-2 -mx-4 overflow-x-auto sm:-mx-6 lg:-mx-8
     grid2Class: "",
     //original: inline-block min-w-full py-2 align-middle md:px-6 lg:px-8
    grid3Class: "inline-block min-w-full py-2 align-middle",
    grid4Class: "overflow-hidden shadow ring-1 ring-black/5 md:rounded-lg",
    tableClass: "min-w-full divide-y divide-gray-200 dark:divide-gray-700",
    theadClass: "bg-gray-50 dark:bg-gray-900",
    tableCellClass: "px-6 py-4 whitespace-nowrap text-sm text-gray-500 dark:text-gray-400",
    theadRowClass: "select-none",
    theadCellClass: "px-6 py-4 text-left text-sm font-medium tracking-wider whitespace-nowrap",
    toolbarButtonClass: "inline-flex items-center px-2.5 py-1.5 border border-gray-300 dark:border-gray-700 shadow-sm text-sm font-medium rounded text-gray-700 dark:text-gray-300 bg-white dark:bg-black hover:bg-gray-50 dark:hover:bg-gray-900 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 dark:ring-offset-black",
}

File: MyApp/Pages/Css.cs
namespace MyApp;

public static class Css
{
    public const string H1 = "text-3xl font-bold tracking-tight text-gray-900 dark:text-gray-50 sm:text-4xl";
    public const string H2 = "text-3xl md:text-4xl font-bold tracking-tighter leading-tight md:pr-8 whitespace-nowrap pt-8";
    public const string H3 = "mb-4 text-2xl font-semibold text-gray-900 dark:text-gray-100 leading-tight";
    public const string H4 = "text-xl font-semibold text-gray-900 dark:text-gray-100 leading-tight";
    public const string Link = "font-semibold text-indigo-600 dark:text-indigo-300 hover:text-indigo-500 dark:hover:text-indigo-400";
    public const string LinkUnderline = "underline hover:text-success duration-200 transition-colors";
    public const string PrimaryButton = "cursor-pointer inline-flex justify-center rounded-md border border-transparent py-2 px-4 text-sm font-medium shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 dark:ring-offset-black focus:ring-2 focus:ring-offset-2 text-white bg-indigo-600 hover:bg-indigo-700 focus:ring-indigo-500 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800";
    public const string SecondaryButton = "cursor-pointer inline-flex justify-center rounded-md border border-gray-300 py-2 px-4 text-sm font-medium shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 bg-white dark:bg-gray-800 border-gray-300 dark:border-gray-600 text-gray-700 dark:text-gray-400 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-700 focus:ring-indigo-500 dark:focus:ring-indigo-600 dark:ring-offset-black";
    public const string DangerButton = "cursor-pointer inline-flex justify-center rounded-md border border-transparent py-2 px-4 text-sm font-medium shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 dark:ring-offset-black focus:ring-red-300 text-white bg-red-600 hover:bg-red-700 focus:ring-red-500 dark:bg-red-600 dark:hover:bg-red-700 dark:focus:ring-red-500 dark:ring-offset-black";
    public const string LabelClasses = "block text-sm font-medium text-gray-700 dark:text-gray-300";
    public static string InputText => CssUtils.Tailwind.Input("");
    public const string InputCheckbox = "focus:ring-indigo-500 h-4 w-4 text-indigo-600 rounded border-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:ring-offset-black";
    public const string AlertDanger = "mb-2 p-2 px-4 flex items-center rounded text-danger bg-red-100 font-semibold";
    public const string AlertSuccess = "mb-2 p-2 px-4 flex items-center rounded text-success bg-green-100 font-semibold";
    public static string ClassNames(params string?[] classes) => CssUtils.ClassNames(classes);
}
